private_chaincode_apps/echo/echo_cc.cpp

```diff
@@ -0,0 +1,261 @@
+/*
+* Copyright 2019 Intel Corporation
+* Copyright IBM Corp. 2018 All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+#include "echo_cc.h"
+#include "echo_json.h"
+#include "logging.h"
+#include "shim.h"
+
+#include <vector>
+
+#define MAX_VALUE_SIZE 1024
+
+#define OK "OK"
+#define AUCTION_DRAW "DRAW"
+#define AUCTION_NO_BIDS "NO_BIDS"
+#define AUCTION_ALREADY_EXISTING "AUCTION_ALREADY_EXISTING"
+#define AUCTION_NOT_EXISTING "AUCTION_NOT_EXISTING"
+#define AUCTION_ALREADY_CLOSED "AUCTION_ALREADY_CLOSED"
+#define AUCTION_STILL_OPEN "AUCTION_STILL_OPEN"
+
+static std::string SEP = ".";
+static std::string PREFIX = SEP + "somePrefix" + SEP;
+
+// implements chaincode logic for invoke
+int invoke(const char* args, uint8_t* response, uint32_t max_response_len,
+    uint32_t* actual_response_len, void* ctx)
+{
+    LOG_DEBUG("EchoCC: +++ Executing chaincode invocation +++");
+    LOG_DEBUG("Args: %s", args);
+
+    std::vector<std::string> argss;
+    // parse json args
+    unmarshal_args(argss, args);
```

Comment:
when I see this, I actually think the shim should provide the args already as array instead of unmarshalling here. I add a note in the shim refactoring issue #83

Reply:
echo has to be cleaned up in a later PR